File: src/ruby/lib/chitti/handler_interceptors/statsd_interceptor.rb
require 'statsd-instrument'

class StatsdMiddleware < ::GRPC::ServerInterceptor
  def request_response(request:, call:, method:)
    service_name = method.receiver.class.respond_to?(:service_name) ? method.receiver.class.service_name : method.receiver.class.name
    begin
      response = StatsD.measure("chitti.#{service_name}.response.time,method=#{method.name}") { yield }
      StatsD.increment("chitti.#{service_name}.success,method=#{method.name}")
      response
    rescue Exception => exception
      StatsD.increment("chitti.#{service_name}.failure,method=#{method.name}")
      raise exception
    end
  end
end

File: src/node-src/handler_interceptors/statsd_interceptor.js
import HandlerInterceptor from '../core/handler_interceptor';
import StatsD from '../lib/statsd';

class StatsDInterceptor extends HandlerInterceptor {
    async call(request, next, ctx) {
        try {
            const response = await StatsD.benchmark(
                next(request),
                `${process.env.SERVICE_NAME || process.env.Service || 'unnamed-service'}.response.time` +
                `,api=${ctx.service}.${ctx.method}`,
            );
            StatsD.increment(`${process.env.SERVICE_NAME || process.env.Service || 'unnamed-service'}.success`
                + `,api=${ctx.service}.${ctx.method}`);
            return response;
        }
        catch (err) {
            StatsD.increment(`${process.env.SERVICE_NAME || process.env.Service || 'unnamed-service'}.failure`
                + `,api=${ctx.service}.${ctx.method}`);
            throw err;
        }
    }
}

export default StatsDInterceptor;
